active/0003-new-config-files.md

```diff
@@ -95,21 +123,142 @@ to
 force_shutdown_policy: 10000|64MB
 ```
 
-The prior is not only more readable to users without reading the comments/documents, but also more intuitive to programmers for searching the config entry through the code files.
+The prior is not only more readable to users without reading the comments/documents, but also more
+intuitive to programmers for searching the config entry through the code files, as the ConfMap
+maintained in the memory is of the same structure (and name) as the config file.
+
+### The Structure of the Config Files
+
+At the beginning of the `emqx.conf` is the notes:
+
+```
+## NOTE: The configurations in this file will be overridden by `/etc/emqx_overrides.conf`
+## NOTE: See the /etc/docs/emqx_config_sample.conf for all the available configurations and
+##       examples.
+```
+
+The first thing to be configured is the node name:
+
+```
+node {
+  name: "emqx@127.0.0.1"
+}
+```
+
+After that is the listeners:
+
+```
+listeners {
```

Comment:
I hope that the listener can define its own mqtt parameters, and the configuration should be organized hierarchically.
For example, `global → protocol → zone → listener → topic`
Configuration items can be overridden from high level to low level.

```
 max_packet_size: 1MB,
 wildcard_subscription: true
 
 zone {
   name: external_zone
   
   # external zone need reset mqtt package size
   max_packet_size: 128KB
   
   listener {
      type: ssl
      port: 8883
      certfile: etc/certs/cert.pem
      keyfile: etc/certs/key.pem
      acceptors: 16
      max_connections: 100000
   }
   
   listener {
      type: tcp
      address: 0.0.0.0
      port: 1883
   }
 }
```

Reply:
As we discussed in the last session, we put the zone out sized the listeners:

```
zones.external : {
  listeners.tcp: {
    bind: …
  }
  mqtt : {
    ...
  }
}
```

Reply:
Is the configuration nesting of MQTT redundant?

all MQTT option in `etc/emqx.conf`:

```
mqtt.max_packet_size = 1MB
mqtt.max_clientid_len = 65535
mqtt.max_topic_levels = 0
mqtt.max_qos_allowed = 2
mqtt.max_topic_alias = 65535
mqtt.retain_available = true
mqtt.wildcard_subscription = true
mqtt.shared_subscription = true
mqtt.ignore_loop_deliver = false
mqtt.strict_mode = false
```

#### Mode 1
This nesting is a little strange，does zone’s `mqtt` full coverage global `mqtt` ?

·```
{ "mqtt": { "a": 1, "b": "1" } }
+
{ "mqtt": { "a": 2 } }
=
{ "mqtt": { "a": 2 } }

"b": 1 will be drop.
```

```js
// javascript

var a = { mqtt: { a: 1, b: 2 } }
var b = { mqtt: { a: 2 } }

// 合并
var c = { ...a, ...b }

// 2
console.log(c.mqtt.a)

// undefined
console.log(c.mqtt.b)
```

Intuitively, objects are nested according to the key.

```
# default value
# mqtt.max_topic_alias = 65535

# global
mqtt: {
   max_pacaket_size: 1024KB
   max_topic_alias = 100
}

zones.external: {
  # zone
  mqtt: {
    max_pacaket_size: 1024KB
  }

  listners.tcp: {
    bind: "0.0.0.0.0",
    port: 1883
    # listner
    mqtt: {
      max_pacaket_size: 1024KB
    }
  }
}
```





#### Mode2

I see some config in `etc/zones.conf`,
I actually know what MQTT configuration items are.

```
zone.internal.max_subscriptions = 0
zone.internal.max_inflight = 128
zone.external.max_packet_size = 1MB
```

so, discard MQTT nesting，`mqtt: {} -> {}`
```
max_pacaket_size: 1024KB

zons.external: {
  max_pacaket_size: 1024KB

  listners.tcp: {
    bind: "0.0.0.0.0",
    port: 1883
    max_pacaket_size: 1024KB
  }
}
```

### Mode3
Adding an `mqtt_` prefix might be a better way because we rarely modify config.

```
mqtt_max_pacaket_size: 1024KB
mqtt_max_qos_allowed: 1024KB

zons.external: {
  mqtt_max_packet_size: 1024KB

  listners.tcp: {
    bind: "0.0.0.0.0",
    port: 1883
    mqtt_max_pacaket_size: 1024KB
  }
}
```

Refer Nginx:
```
# ssl
ssl_certificate         /etc/letsencrypt/live/emqx.io/fullchain.pem;
ssl_certificate_key     /etc/letsencrypt/live/emqx.io/privkey.pem;
ssl_trusted_certificate /etc/letsencrypt/live/emqx.io/chain.pem;

# gzip
gzip                    on;
gzip_vary               on;
gzip_proxied            any;
gzip_comp_level         6;


# Proxy timeouts
proxy_connect_timeout              60s;
proxy_send_timeout                 60s;
proxy_read_timeout                 60s;


# logging
access_log /var/log/nginx/emqx.io.access.log;
error_log  /var/log/nginx/emqx.io.error.log warn;
```